### PR TITLE
[Feature] Modal for Import in Example Submission Board

### PR DESCRIPTION
**Pull request by imstreetspirit**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-testing/).
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all newly inserted strings into English and German.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).

### Motivation and Context

Follow-up PR for #3931 

This PR introduces new modal for importing example submission. It enables to instructor to see all student submissions are ready to import. That means the submission should be assessed by tutor/instructor to see it in the list. Only instructor can decide submission to be imported. Afterwards, this submission can be used as an example submission to train tutors. Submissions can be searched by participant name in the modal.

### Description

Added pageable query in `StudentParticipantRepository`. After retrieving participants for the exercise, it finds latest submission of student and makes pageable response for client side in `SubmissionService`. 

added `example-submission-import.component` and respective html template and spec file. It displays submissions in table. Also, uses `example-submission-import-paging.service` to retrieve the pageable submission content from server. and respective spec file added for paging service. 

added Import button in `example-submission.component`. It opens Import modal. To import the submission, it uses `exampleSubmissionService ` and sends the selected submission id and exercise id after closing the modal.

### Steps for Testing

1. Log in to Artemis with Instructor account
2. Navigate to Course Management
3. Find  or create Text/Modeling Exercise
4. Participate to this exercise with student account
5. Assess the student submission as tutor/instructor
6. Login as Instructor
7. Go to Example Submission board for the exercise
8. Click Import button
9. Confirm that you can see the submission in the table
10. Click Import button for that submission
11. Confirm that page is re-loaded and the new example submission is on the list
12. Then open the example submission
13. Confirm that submission is imported properly with its assessment

Plus: You can add more than one student submission for the exercise to see the submission list is retrieved properly when you open import modal for example submission.

### Review Progress

- Code Review
  - [x] Review 1
  - [x] Review 2
- Manual Tests
  - [x] Test 1
  - [x] Test 2

### Test Coverage

(only covered new implementation)
SubmissionResource.java 92%
SubmissionService.java 93%
example-submission-import-paging.service.ts 100%
example-submission-import.component.ts 96%

### Screenshots

Example Submission Board

Modal for Example Submission Import

### CODE REVIEW

File: src/main/webapp/app/exercises/shared/example-submission/example-submission-import/example-submission-import.component.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    get sortedColumn(): string {
+        return this.state.sortedColumn;
+    }
+
+    /**
+     * Set the list sorting direction
+     *
+     * @param ascending {boolean} Ascending order set
```
Comment (lschlesinger) [inline src/main/webapp/app/exercises/shared/example-submission/example-submission-import/example-submission-import.component.ts:124]: maybe the comment "flag indicating if the sorting order is set to ascending or not" is easier to understand?

**Review by lschlesinger — CHANGES_REQUESTED** [review]
It's a great feature, I added some minor comments and have some general questions/remarks:
1. When I import an example submission, I think it would be helpful to indicate, that the search term is related to the student name. Now with "Search for Submission [Input]", I would not know if I have to provide an Id, or a date or whatever... it's just more clear, if you write: "Search for Submission by Student Name", because this is, what is actually done right?
2. When I successfully imported an example submission, I afterwards have no clear indication, which of the shown submissions in the table is the one I imported. As you cannot preview the submission before selecting it for import, I can imagine, that after the import, I want to reassure myself, that the submission is correctly imported and maybe edit it. So I have to find the imported solution in the table, but the title is generated automatically, the ID that is shown in the table was not visible during the selection in the modal. I could use the count that is used within the title, but maybe you'll find another solution to indicate after modal closing, which of the submissions is the imported one (if you think it's worth improving it, this can also be done later on).
3. Regarding consistency, I think the buttons for creating and importing a submission could be consistent with the buttons to import and create an exercise. Having the '+' only for the import button is not perfect, as the create button also adds (+) solutions. The exercise buttons have the '+' for both buttons, see images:
Bildschirmfoto 2021-09-13 um 12 25 26
Bildschirmfoto 2021-09-13 um 12 25 18

4. The modal should also close when I navigate back via browser, I think this is achieved by invoking the `.close()` method on the `modalRef`  in the destroy lifecycle hook of the component, the modal is living in. I had the same problem, I implemented the suggested solution[ here](https://github.com/ls1intum/Artemis/pull/3934/files#diff-299a30610e5541cf3ace01c07d4264554e4176d6c25bcde81903bcd37e361ab8).
This is the behavior right now:

https://user-images.githubusercontent.com/41366522/133070724-344265df-a558-460c-a1b8-4c6384e3831b.mov

**Comment by imstreetspirit** [comment]
Thank you for the review and testing effort 😊 

> 1. When I import an example submission, I think it would be helpful to indicate, that the search term is related to the student name. Now with "Search for Submission [Input]", I would not know if I have to provide an Id, or a date or whatever... it's just more clear, if you write: "Search for Submission by Student Name", because this is, what is actually done right?

yes, search functionality is working with student name. you're right. I will indicate it exactly in UI. What is the correct translation for "Search for Submission by Student Name". what do you think about this one "Suche nach Einreichungen anhand des Namens des Studierende:r" 😄 

> 2. When I successfully imported an example submission, I afterwards have no clear indication, which of the shown submissions in the table is the one I imported. As you cannot preview the submission before selecting it for import, I can imagine, that after the import, I want to reassure myself, that the submission is correctly imported and maybe edit it. So I have to find the imported solution in the table, but the title is generated automatically, the ID that is shown in the table was not visible during the selection in the modal. I could use the count that is used within the title, but maybe you'll find another solution to indicate after modal closing, which of the submissions is the imported one (if you think it's worth improving it, this can also be done later on).

ohh, I see what you meant. it's really good point. I'll think about that. currently title column is automatically filled with the index for example-submission-table. agree with you, It would be great to know which one is the imported one in the example submission board.

> 3. Regarding consistency, I think the buttons for creating and importing a submission could be consistent with the buttons to import and create an exercise. Having the '+' only for the import button is not perfect, as the create button also adds (+) solutions. The exercise buttons have the '+' for both buttons, see images:

oh, right, I will add plus icon for the create button. thanks for catching it.

> 1. The modal should also close when I navigate back via browser, I think this is achieved by invoking the `.close()` method on the `modalRef`  in the destroy lifecycle hook of the component, the modal is living in. I had the same problem, I implemented the suggested solution[ here](https://github.com/ls1intum/Artemis/pull/3934/files#diff-299a30610e5541cf3ace01c07d4264554e4176d6c25bcde81903bcd37e361ab8).

ohh, did not know that. thanks for the suggestion. I'll take care of destroying/unmounting the component.

File: src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java [hunk 2]
```diff
@@ -405,6 +407,24 @@ SELECT max(id) FROM p.results)
             """)
     List<StudentParticipation> findAllWithEagerSubmissionsAndEagerResultsAndEagerAssessorByExerciseIdIgnoreTestRuns(@Param("exerciseId") long exerciseId);
 
+    @Query(value = """
+            SELECT p FROM StudentParticipation p
+            LEFT JOIN FETCH p.submissions s
+            LEFT JOIN FETCH p.results r
+            WHERE p.exercise.id = :#{#exerciseId}
+                  AND (p.student.firstName LIKE %:partialStudentName% OR p.student.lastName LIKE %:partialStudentName%)
```
[the remark this answers is not in the document]
Comment (imstreetspirit) [reply src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java:415]: this is for search functionality in UI. when you type student name, query gets the search term as `partialStudentName`, then compare with first name or last name

[the remark this answers is not in the document]
Comment (lschlesinger) [reply src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java:415]: ah yes, I saw that during manual test. When I search for "Anna", all the participations of a student with name "Anna" appear, so if there are several Annas, I can still select the correct one when I browse the list of found participation. We do not rely on the perfect match in that query...I understand that now :)

File: src/main/java/de/tum/in/www1/artemis/service/SubmissionService.java [hunk 3]
```diff
@@ -641,4 +644,36 @@ private void prepareComplaintAndSubmission(Complaint complaint, Submission submi
...
+     * Search for all submissions fitting a {@link PageableSearchDTO search query}. The result is paged,
+     * meaning that there is only a predefined portion of the result returned to the user, so that the server doesn't
+     * have to send hundreds/thousands of submissions if there are that many in Artemis.
+     *
+     * @param search     The search query defining the search term and the size of the returned page
+     * @param exerciseId exerciseId which submissions belongs to
+     * @return A wrapper object containing a list of all found submissions and the total number of pages
+     */
+    public SearchResultPageDTO<Submission> getSubmissionsOnPageWithSize(PageableSearchDTO<String> search, Long exerciseId) {
+        var sorting = Sort.by(StudentParticipation.StudentParticipationSearchColumn.valueOf(search.getSortedColumn()).getMappedColumnName());
```
[the remark this answers is not in the document]
Comment (imstreetspirit) [reply src/main/java/de/tum/in/www1/artemis/service/SubmissionService.java:658]: ah, okay. I'll look at it. I took this part from Import exercise function, but yeah I'm not the fan of using `var`

File: src/main/webapp/app/exercises/shared/example-submission/example-submission-import/example-submission-import-paging.service.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+type EntityResponseType = SearchResult<Submission>;
+
+@Injectable({ providedIn: 'root' })
+export class ExampleSubmissionImportPagingService {
+    constructor(private http: HttpClient) {}
+
+    /**
+     * Gets all submissions with exerciseId
+     * @param pageable   pageable search containing the page size and query string
```
[the remark this answers is not in the document]
Comment (imstreetspirit) [reply src/main/webapp/app/exercises/shared/example-submission/example-submission-import/example-submission-import-paging.service.ts:16]: I'll make it explicit

**Comment by lschlesinger** [comment]
Regarding the translation... it's a bit tricky, it should not be too long I guess.
"Suche nach Einreichungen [INPUT] (Name d. Studierenden)" (avoid to distinguish between _der_ Studierenden (f) and _des_ Studierenden (m), which is a bit ugly)
or 
"Suche nach Einreichungen [Name Studierende:r]" so using the placeholder instead (my preferred solution :))
or
"Suche nach Einreichungen mit Studierenden Name [ INPUT FIELD]"

**Comment by imstreetspirit** [comment]
> Regarding the translation... it's a bit tricky, it should not be too long I guess.
> "Suche nach Einreichungen [INPUT] (Name d. Studierenden)" (avoid to distinguish between _der_ Studierenden (f) and _des_ Studierenden (m), which is a bit ugly)
> or
> "Suche nach Einreichungen [Name Studierende:r]" so using the placeholder instead (my preferred solution :))
> or
> "Suche nach Einreichungen mit Studierenden Name [ INPUT FIELD]"

I liked the second one, others look a little bit long 😊

**Review by Olimbia23 — DISMISSED** [review]
I like the feature @imstreetspirit 😀 
Tested on TS1 during testing session. The student submission together with the assessment are imported successfully. 
Will check the code another day as well. 😊

**Review by Malyuk-A — APPROVED** [review]
Tested on TS1. Works as described 👍
Code also looks good.

**Review by ndricimrr — APPROVED** [review]
Tested the functoinality on TS1. Worked as described for text exercises. Couldn't spot any issues.

**Review by argertboja — DISMISSED** [review]
Great feature and nicely implemented 👍🏼 Tested on TS1 and encountered no issues!

File: src/main/webapp/app/exercises/shared/example-submission/example-submissions.component.ts [hunk 5]
```diff
@@ -69,4 +83,28 @@ export class ExampleSubmissionsComponent implements OnInit {
...
+            (selectedSubmission: Submission) => {
+                this.exampleSubmissionService.import(selectedSubmission.id!, this.exercise.id!).subscribe(
+                    () => {
+                        this.alertService.success('artemisApp.exampleSubmission.submitSuccessful');
+                        location.reload();
+                    },
+                    (error: HttpErrorResponse) => onError(this.alertService, error),
+                );
+            },
+            () => {},
```
Comment (lschlesinger) [inline src/main/webapp/app/exercises/shared/example-submission/example-submissions.component.ts:107]: I think,`()=>{}` is not necessary.

**Review by lschlesinger — APPROVED** [review]
I approve the code, you've addressed all the changes. I had a minor suggestion, which you can try out.

File: src/main/webapp/app/exercises/shared/example-submission/example-submissions.component.ts (hunk 5 above)
Comment (imstreetspirit) [reply src/main/webapp/app/exercises/shared/example-submission/example-submissions.component.ts:107]: additionally, I do not know how to make Client Test green 😢

**Review by lschlesinger — APPROVED** [review]
Approve the code after minor recent changes 👍

File: src/main/webapp/app/exercises/shared/example-submission/example-submissions.component.ts [hunk 6]
```diff
@@ -69,4 +83,25 @@ export class ExampleSubmissionsComponent implements OnInit {
...
+        const exampleSubmissionImportModalRef = this.modalService.open(ExampleSubmissionImportComponent, {
+            size: 'lg',
+            backdrop: 'static',
+        });
+        exampleSubmissionImportModalRef.componentInstance.exercise = this.exercise;
+        exampleSubmissionImportModalRef.result.then((selectedSubmission: Submission) => {
+            this.exampleSubmissionService.import(selectedSubmission.id!, this.exercise.id!).subscribe(
+                () => {
+                    this.alertService.success('artemisApp.exampleSubmission.submitSuccessful');
+                    location.reload();
```
Comment (MDK5) [inline src/main/webapp/app/exercises/shared/example-submission/example-submissions.component.ts:101]:
```suggestion
                    this.ngOnInit();
```
might also work, according to https://stackoverflow.com/a/51453244/11948496

**Review by MDK5 — APPROVED** [review]
Tested in testing session on ts2, worked as described

**Review by willergeorg — APPROVED** [review]
Tested during testing session. No issues found.

**Review by niclasschuemann — APPROVED** [review]
Tested in testing session, good job 👍

**Review by Malyuk-A — APPROVED** [review]
Tested during testing session on TS2. re-approve

**Review by coolchock — APPROVED** [review]
tested in testing session, works as expected

**Review by ndricimrr — APPROVED** [review]
Tested on Testing Session. No issues found.

**Review by cremertim — APPROVED** [review]
tested during testing session

**Pull request merged**